src/crunch/crn_decomp.rs

```diff
@@ -0,0 +1,347 @@
+        let res: bool = pHeader.crnd_get_header(pData, data_size); 
+        if res == false {
+            return res;
+        }
+
+        self.m_width = pHeader.m_width.cast_to_uint();
+        self.m_height = pHeader.m_height.cast_to_uint();
+        self.m_levels = pHeader.m_levels.cast_to_uint();
+        self.m_faces = pHeader.m_faces.cast_to_uint();
+        self.m_format = match pHeader.m_format.cast_to_uint(){
+            // -1 => crn_format::cCRNFmtInvalid,
+
+            0 => crn_format::cCRNFmtDXT1,
+        
+            // 0 => crn_formatcCRNFmtFirstValid,
+        
+            // cCRNFmtDXT3 is not currently supported when writing to CRN - only DDS.
+            1 => crn_format::cCRNFmtDXT3,
+        
+            2 => crn_format::cCRNFmtDXT5,
+        
+            // Various DXT5 derivatives
+            3 => crn_format::cCRNFmtDXT5_CCxY,    // Luma-chroma
+            4 => crn_format::cCRNFmtDXT5_xGxR,    // Swizzled 2-component
+            5 => crn_format::cCRNFmtDXT5_xGBR,    // Swizzled 3-component
+            6 => crn_format::cCRNFmtDXT5_AGBR,    // Swizzled 4-component
+        
+            // ATI 3DC and X360 DXN
+            7 => crn_format::cCRNFmtDXN_XY,
+            8 => crn_format::cCRNFmtDXN_YX,
+        
+            // DXT5 alpha blocks only
+            9 => crn_format::cCRNFmtDXT5A,
+        
+            10 => crn_format::cCRNFmtETC1,
+        
+            11 => crn_format::cCRNFmtTotal,
+        
+            0xFFFFFFFF => crn_format::cCRNFmtForceDWORD,
+
+            _ => crn_format::cCRNFmtInvalid
+        };
+        if self.m_format == crn_format::cCRNFmtInvalid {
+            return false;
+        }
+        if (pHeader.m_format.cast_to_uint() == crn_format::cCRNFmtDXT1 as u32) || (pHeader.m_format.cast_to_uint() == crn_format::cCRNFmtDXT5A as u32) {
+            self.m_bytes_per_block = 8;
+        }else{
+            self.m_bytes_per_block = 16;
+        }
+        self.m_userdata0 = pHeader.m_userdata0.cast_to_uint();
+        self.m_userdata1 = pHeader.m_userdata1.cast_to_uint();
+        return true;
+    }
+}
+
+#[repr(C)]
+pub struct crn_packed_uint<const N: usize>{
+    m_buf: [u8; N]
+}
+
+// no-std, so we can not use std::ops
+impl<const N: usize> crn_packed_uint<N>{
+    pub fn assign(&mut self, other: crn_packed_uint<N>) -> &crn_packed_uint<N>{
+        if self.m_buf != other.m_buf{ // this works?
+            self.m_buf.copy_from_slice(&other.m_buf[0..N]);
+        }
+        return self;
+    }
+
+    pub fn assign_from_buffer(&mut self, other: &[u8]){
+        self.m_buf.copy_from_slice(&other[0..N])
+    }
+
+    // Rust doesn't support function overloading.
+    pub fn assign_val(&mut self, mut val: u32) -> &crn_packed_uint<N>{
+        val <<= 8 * (4 - N);
+        for i in 0..N{
+            self.m_buf[i] = (val >> 24) as u8;
+            val <<= 8; 
+        }
+        return self;
+    }
+
+    pub fn cast_to_uint(&mut self) -> u32{
+        return match N {
```

Comment:
This could be replaced with

```rust
 `        return match N {
            1 => self.m_buf[0] as u32,
            2 => u16::from_be_bytes(self.m_buf),
            3 => self.m_buf[0] << 16 | u16::from_be_bytes(self.m_buf[1..=2]),
            4 => u32::from_be_bytes(self.m_buf)
        }
        ````
        to make it easier to read

Reply:
and return keyword should be dropped